### PR TITLE
feat: Support Top-K logprobs and prompt_logprobs in LLMAPI

### PR DESCRIPTION
**Pull request by hchings**
This MR only handles TRT path. PyT flow will be handling in [this MR](https://github.com/NVIDIA/TensorRT-LLM/pull/3221/).

## Usage
```python

llm = LLM(
    model="TinyLlama/TinyLlama-1.1B-Chat-v1.0",
    build_config=BuildConfig(gather_context_logits=True),
    gather_generation_logits=True)
        
sampling_params = SamplingParams(
        temperature=0.8,
        top_p=0.95,
        logprobs=5,   # new
        prompt_logprobs=2 # new
        ) 
        
# output.outputs[0].logprobs (list of dict, where the dict is token_id to Logprob): 
# [{11785: Logprob(logprob=-2.128943920135498, rank=1), 1244: Logprob(logprob=-2.878943920135498, rank=2), 297: Logprob(logprob=-3.003943920135498, rank=3), 451: Logprob(logprob=-3.316443920135498, rank=4), 263: Logprob(logprob=-3.503943920135498, rank=5), 2086: Logprob(logprob=-6.722693920135498, rank=113)}, {17999: Logprob(logprob=-1.5640053749084473, rank=1), 11785: Logprob(logprob=-1.7515053749084473, rank=2), 4280: Logprob(logprob=-2.5015053749084473, rank=3), 4100: Logprob(logprob=-3.0640053749084473, rank=4), 5566: Logprob(logprob=-3.1890053749084473, rank=5)}, {304: Logprob(logprob=-0.16267399489879608, rank=1), 363: Logprob(logprob=-2.6626739501953125, rank=2), 29892: Logprob(logprob=-4.2876739501953125, rank=3), 29889: Logprob(logprob=-4.3501739501953125, rank=4), 322: Logprob(logprob=-4.6001739501953125, rank=5)}...]       


# output.outputs[0].prompt_logprobs
#     same shape as above


 
```

### CODE REVIEW

File: tensorrt_llm/llmapi/llm.py [hunk 1]
```diff
@@ -447,6 +447,12 @@ def _prepare_sampling_params(
                         "tokenizer is required to reset end_id if it is None, or you can explicitly specify the end_id for sampling_params"
                     )
                 sampling_params._setup(self.tokenizer)
+            if sampling_params.prompt_logprobs:
+                # Context logits are required to compute prompt_logprobs
+                sampling_params.return_context_logits = True
+                # TODO: need a way to know whether to drop context logits at worker later
```
Comment (hchings) [inline tensorrt_llm/llmapi/llm.py]: @Superjomn  - for me to dump contexts (or generation logits) before sending it via IPC, I need a flag to know whether users actually set `return_context_logits`  or it was enabled here.  

It could be achieved by adding an additional internal flag in `SamplingParams`, lmk if you suggest against that.

File: tensorrt_llm/executor/worker.py [hunk 2]
```diff
@@ -747,16 +747,27 @@ def handle_for_worker(self, responses: List[tllm.Response]) -> None:
             assert response is not None
             queue = self.worker.return_queue(response.client_id)
 
+            # TODO: pass in k
```
Comment (hchings) [inline tensorrt_llm/executor/worker.py]: @Superjomn  - I need to pass in `k` (aka `logprobs` or `prompt_logprobs`) from `request.sampling_params` to here.  Haven't found a clean way given how worker is currently structured.

File: tensorrt_llm/executor/result.py [hunk 3]
```diff
@@ -532,3 +565,58 @@ async def __anext__(self):
         except asyncio.TimeoutError:
             self._done = True
             raise StopAsyncIteration
+
+
+def compute_logprobs(
```
Comment (hchings) [inline tensorrt_llm/executor/result.py:594]: I put all logprob stuff in `result.py`, we can move to a single file if needed.

File: tensorrt_llm/sampling_params.py [hunk 4]
```diff
@@ -280,13 +282,6 @@ def __post_init__(self):
 
         self.best_of = self.best_of or self.n
 
-        if (not self.use_beam_search and self.n < self.best_of
```
Comment (hchings) [inline tensorrt_llm/sampling_params.py:283]: I'm not exactly sure what this is, still confirming w/ Xiaowei this is no longer needed.

File: tensorrt_llm/executor/result.py [hunk 5]
```diff
@@ -29,6 +31,17 @@
 ]
 
 
+@dataclass
+class Logprob:
```
Comment (Superjomn) [inline tensorrt_llm/executor/result.py:36]: `class Logprob(slots=True)` is better for performance.

File: tensorrt_llm/executor/result.py [hunk 6]
```diff
@@ -76,6 +91,8 @@ def __post_init__(self):
             self.token_ids = []
         if self.logprobs is None:
             self.logprobs = []
+        if self.prompt_logprobs is None:
+            self.prompt_logprobs = []
```
Comment (Superjomn) [inline tensorrt_llm/executor/result.py]: For setting default values with containers, we can declare the parameters in dataclass as
```python
logprobs: LogprobResults = field(default_factory=list)
```
This eliminates the need for the code here.

So are other parameters here.

File: tensorrt_llm/executor/result.py [hunk 7]
```diff
@@ -90,7 +107,7 @@ def token_ids_diff(self) -> List[int]:
         return self.token_ids[self._last_token_ids_len:]
 
     @property
-    def logprobs_diff(self) -> List[float]:
+    def logprobs_diff(self) -> List[dict[int, Logprob]]:
```
Comment (Superjomn) [inline tensorrt_llm/executor/result.py]: We can unify all the `List/Dict` to `list/dict` as they are suggested from [python 3.9](https://docs.python.org/3.9/library/typing.html#typing.List)

File: tensorrt_llm/executor/result.py [hunk 8]
```diff
@@ -224,8 +244,10 @@ def _handle_sequence(self, finish_reasons, response_tensors,
                     f"Unknown finish reason: {finish_reasons[src_idx]}")
 
     @nvtx_range("handle_response", color="red", category="GenerationResultBase")
-    def _handle_response(self, response: Union["PostprocWorker.Output",
-                                               tllm.Response, ErrorResponse]):
+    def _handle_response(self,
+                         response: Union["PostprocWorker.Output", tllm.Response,
+                                         ErrorResponse],
+                         additional_outputs: Optional[Dict[str, Any]] = None):
```
Comment (Superjomn) [inline tensorrt_llm/executor/result.py]: It seems that the addtional_outputs contains multiple members, and Dict is an opaque compared to a concrete type. How about making it a namedtuple:
```python
class AdditonalOutputs(Namedtuple):
    ''' some comments '''
    prompt_logprobs: Optional[...] = None
    generation_logprobs: Optional[...] = None
```

This can make this argument more intuitive and good for jumping in IDE as well.

File: tensorrt_llm/executor/result.py [hunk 9]
```diff
@@ -532,3 +565,58 @@ async def __anext__(self):
...
             raise StopAsyncIteration
+
+
+def compute_logprobs(
+    context_logits: Optional[torch.Tensor],
+    generation_logits: Optional[torch.Tensor],
+    output_token_ids: Optional[list[int]],
+    k_prompt_logprobs: int = 5,
+    k_logprobs: int = 5,
+) -> Tuple[Optional[LogprobResults], Optional[LogprobResults]]:
```
Comment (Superjomn) [inline tensorrt_llm/executor/result.py]: replace Tuple => tuple.

File: tensorrt_llm/executor/result.py [hunk 10]
```diff
@@ -532,3 +565,58 @@ async def __anext__(self):
...
+
+    Returns:
+        Tuple of:
+            - prompt_logprobs: list of dicts {token_id -> Logprob}
+            - generation_logprobs: list of dicts {token_id -> Logprob}
+    """
+
+    def _topk_logprobs(logits: torch.Tensor, top_k: int,
+                       tokens: Optional[list[int]]) -> LogprobResults:
+        if logits.dim() == 3 and logits.size(0) == 1:
```
Comment (Superjomn) [inline tensorrt_llm/executor/result.py]: Where do these magic numbers come from? Can we eliminate them by deducing from some shape-like variable, or maybe some comment is helpful for understanding?

File: tensorrt_llm/executor/result.py [hunk 11]
```diff
@@ -532,3 +565,58 @@ async def __anext__(self):
...
+            logits = logits.squeeze(0)  # shape: [T, vocab]
+
+        logprobs = F.log_softmax(logits.to("cuda", dtype=torch.float32), dim=-1)
```
Comment (Superjomn) [inline tensorrt_llm/executor/result.py:616]: Please add a `NOTE` here to mark the potential performance drop, as we may reuse this part and enhance in PyT sampler.

File: tensorrt_llm/executor/result.py [hunk 12]
```diff
@@ -532,3 +565,58 @@ async def __anext__(self):
...
+        topk_vals, topk_indices = torch.topk(logprobs, k=top_k, dim=-1)
+
+        results: Logprobs = []
```
Comment (Superjomn) [inline tensorrt_llm/executor/result.py]: It seems the Logprobs is not bound; should it be `list[Logprob]`?

File: tensorrt_llm/executor/worker.py (hunk 2 above)
Comment (Superjomn) [reply tensorrt_llm/executor/worker.py]: Can you elaborate on this?

File: tensorrt_llm/executor/worker.py [hunk 13]
```diff
@@ -747,16 +747,27 @@ def handle_for_worker(self, responses: List[tllm.Response]) -> None:
...
+                "prompt_logprobs": prompt_logprobs,
+                "generation_logprobs": logprobs,
+            }
+
             # For AsyncQueue.sync_q, we will batch the events to avoid too many
             # event notifications, thus put without wait here.
             if isinstance(queue, _SyncQueue):
                 global_tracer().log_instant("worker-rsp.put")
-                queue.put_nowait(response)
+                queue.put_nowait((response, additional_outputs))
```
Comment (Superjomn) [inline tensorrt_llm/executor/worker.py]: How about making a Namedtuple for this so that we can add the typing for related variables? This could be better for reading and IDE jumping.

File: tensorrt_llm/llmapi/llm.py (hunk 1 above)
Comment (Superjomn) [reply tensorrt_llm/llmapi/llm.py]: Yeah, that makes sense. How about introducing it in SamplingParams as a private property? As SamplingParams is a 1.0 committed API and currently, this implementation is just a WAR.
```python
@property
def _need_return_context_logits(self) -> bool:
    return self.prompt_logprobs
```
This shouldn't break the API stability check.

File: tensorrt_llm/executor/result.py (hunk 10 above)
Comment (hchings) [reply tensorrt_llm/executor/result.py]: Yes, ignore this if clause now, I'm updating this part.

File: tensorrt_llm/executor/result.py (hunk 11 above)
Comment (hchings) [reply tensorrt_llm/executor/result.py:616]: Do you mean perf drop due to a few more additional operation?

Comment (hchings) [reply tensorrt_llm/executor/result.py:590]: Do you mean perf drop due to a few more additional operation?

File: tensorrt_llm/executor/worker.py (hunk 2 above)
Comment (hchings) [reply tensorrt_llm/executor/worker.py]: Please see changes here: https://github.com/NVIDIA/TensorRT-LLM/pull/3388/files#diff-1b6c3b1327b4cb6f697a78ad27820f044fd5ee2b2738c8851d680b8b8a0d8a7fR431

Basically need to pass some logprob-related sampling_params to `handle_response()`. Not sure whether there's a cleaner method than above link.

File: tensorrt_llm/executor/worker.py [hunk 14]
```diff
@@ -747,16 +764,35 @@ def handle_for_worker(self, responses: List[tllm.Response]) -> None:
...
+                response.request_id)
+            if logprob_params:
+                additional_outputs = compute_logprobs(
+                    logprob_params.get("prompt_logprobs"),
+                    logprob_params.get("logprobs"),
+                    response.result.context_logits,
+                    response.result.generation_logits,
+                    response.result.output_token_ids[0])  # TODO
+
+                if logprob_params.get("drop_context_logits"):
```
Comment (hchings) [inline tensorrt_llm/executor/worker.py]: Hi @Superjomn  - to drop logits as we discussed, I'll need to make the `Result` pybind ( `.def_property_readonly("result", &tle::Response::getResult)` in `request.cpp` ) mutable. 

Can you confirm that there's no concerns of doing so?

File: tensorrt_llm/executor/result.py (hunk 11 above)
Comment (Superjomn) [reply tensorrt_llm/executor/result.py:616]: I mean that the copy the logits tensor from host to cuda, we can record the reason here.

File: tensorrt_llm/executor/worker.py (hunk 14 above)
Comment (Superjomn) [reply tensorrt_llm/executor/worker.py]: I think it is reasonable, but we can leave some comments there to describe the context. In the long run,  we will have some equivalent Python DS, and will unify the DS throughout PyExecutor and LLM-API, so it is a normal need to post-process the Response in each stage.

cc @MartinMarciniszyn @QiJune @kaiyux for viz.

File: tensorrt_llm/executor/worker.py [hunk 15]
```diff
@@ -426,6 +426,19 @@ def submit(self, request: GenerationRequest) -> GenerationResult:
         request_id = self._enqueue_request(request)
         # request_id returned from backend is necessary for the abort_request method.
         self._client_id_to_request_id[client_id] = request_id
+        # Store lobprobs-related fields from request for the later logprob calculation.
+        if request.sampling_params.logprobs or request.sampling_params.prompt_logprobs:
+            self._await_response_helper._request_id_to_logprob_params[
```
Comment (Superjomn) [inline tensorrt_llm/executor/worker.py]: There is one mapping from request_id (actually the client_id) to some DS in worker and proxy, [Worker._results](https://github.com/nvidia/TensorRT-LLM/blob/510c5952cb373d1f0ae21b0fe19190770e2ad1d1/tensorrt_llm/executor/worker.py#L72) and [Proxy.results](https://github.com/nvidia/TensorRT-LLM/blob/510c5952cb373d1f0ae21b0fe19190770e2ad1d1/tensorrt_llm/executor/proxy.py#L79), maybe we can extend the GenerationResult for these new fields.

File: tensorrt_llm/executor/worker.py (hunk 14 above)
Comment (hchings) [reply tensorrt_llm/executor/worker.py]: updated, kept `Result` readonly.

**Comment by MartinMarciniszyn** [comment]
@Funatiq , could you please also take a look at these changes?

File: tensorrt_llm/executor/result.py (hunk 11 above)
Comment (Funatiq) [reply tensorrt_llm/executor/result.py:616]: I'm wondering if we shouldn't do the calculations while the logits are still on the GPU in the model instead of copying them to host and back.

File: tensorrt_llm/executor/worker.py [hunk 16]
```diff
@@ -746,16 +763,33 @@ def handle_for_worker(self, responses: List[tllm.Response]) -> None:
...
+
+                if logprob_params.get("drop_generation_logits"):
+                    response.clear_generation_logits()
+
             # For AsyncQueue.sync_q, we will batch the events to avoid too many
             # event notifications, thus put without wait here.
             if isinstance(queue, _SyncQueue):
                 global_tracer().log_instant("worker-rsp.put")
-                queue.put_nowait(response)
+                queue.put_nowait((response, additional_outputs))
```
Comment (Funatiq) [inline tensorrt_llm/executor/worker.py]: Is it possible to add the `additional_outputs` to `tllm.Result` instead of keeping them separate? There is already a member for it.

**Comment by Funatiq** [comment]
In general, do we consider how users can easily add new outputs to a model? In the TRT workflow we enable additional outputs via the [ExecutorConfig](https://github.com/NVIDIA/TensorRT-LLM/blob/main/cpp/include/tensorrt_llm/executor/executor.h#L1538) and [OutputConfig](https://github.com/NVIDIA/TensorRT-LLM/blob/main/cpp/include/tensorrt_llm/executor/executor.h#L237) and the runtime handles all the rest. Can we implement something similar for the PyTorch workflow?

File: tensorrt_llm/executor/worker.py (hunk 16 above)
Comment (Superjomn) [reply tensorrt_llm/executor/worker.py]: @Funatiq @hchings   Since we are going to have a pure Python Response to replace the pybinding one, keep some new fields in Python and hesitate to introduce new fields in `tllm.Response` if not exist seems acceptable as an intermediate status. But we hope the code to be intuitive enough. One alternative is to make the tuple here a Namedtuple Response with `response` and fields in `additioal_outputs` expanded there with some comments; this way will be better.

Comment (Funatiq) [reply tensorrt_llm/executor/worker.py]: Fine with me.

Comment (hchings) [reply tensorrt_llm/executor/worker.py]: I'll prefer we wait for a pure Python Response, instead of introducing a new Namedtuple (which doesn't differ much than the current impl) and removing it again later.

File: tensorrt_llm/executor/worker.py (hunk 15 above)
Comment (hchings) [reply tensorrt_llm/executor/worker.py]: updated to use `GenerationResult` instead.

File: tensorrt_llm/executor/result.py (hunk 10 above)
Comment (hchings) [reply tensorrt_llm/executor/result.py]: updated

File: tensorrt_llm/executor/executor.py [hunk 17]
```diff
@@ -204,6 +204,23 @@ def _get_next_client_id(self):
         self._last_client_id = (self._last_client_id + 1) & ((1 << 64) - 1)
         return self._last_client_id
 
+    def _get_logprob_params(self, request: GenerationRequest) -> Optional[dict]:
```
Comment (Superjomn) [inline tensorrt_llm/executor/executor.py:207]: A dedicated type for the logprob_params is preferred to help track this DS used across the codebase:

```python
class LogprobParams(Namedtuple):
    logprobs: Logprobs
    prompt_probs: Logprobs
    # Drop the logits once the logprobs are computed from the Worker
    drop_context_logits: bool = False
    # Drop the geneation_logits once the logprobs are computed from the Worker
    drop_generation_logits: bool = False
```

File: tensorrt_llm/sampling_params.py [hunk 18]
```diff
@@ -224,13 +225,16 @@ class SamplingParams:
...
-    return_log_probs: bool = False
+    logprobs: Optional[int] = None
+    prompt_logprobs: Optional[int] = None
     return_context_logits: bool = False
     return_generation_logits: bool = False
     exclude_input_from_output: bool = True
     return_encoder_output: bool = False
     return_perf_metrics: bool = False
     additional_model_outputs: Optional[List[AdditionalModelOutput]] = None
+    _context_logits_auto_enabled: bool = False
```
Comment (Superjomn) [inline tensorrt_llm/sampling_params.py:236]: Can you add some comments here to record the background for the WAR? Since this is TRT-only and SamplingParams is shared between both TRT and PyT, we may deprecate it in the future.

Comment (hchings) [reply tensorrt_llm/sampling_params.py:248]: updated

File: tensorrt_llm/executor/executor.py (hunk 17 above)
Comment (hchings) [reply tensorrt_llm/executor/executor.py]: updated

**Review by Superjomn — APPROVED** [review]
LGTM

**Comment by hchings** [comment]
/bot skip --comment "Latest full pipeline has passed, but CI failed to report back status. The rebase after CI passed is only for sign-off".
